src/coroutine.h

```diff
@@ -0,0 +1,265 @@
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <coroutine>
+#include <type_traits>
+#include <utility>
+#include <variant>
+
+namespace triton::perfanalyzer {
+
+/**
+ * @brief A C++20 coroutine implementation that supports both void and
+ * value-returning coroutines
+ *
+ * @details This class implements a coroutine that can be used to create
+ * cooperative multitasking functionality. It supports both void coroutines and
+ * coroutines that return a value of type T.
+ *
+ * The coroutine starts in a suspended state and must be explicitly resumed
+ * using Resume(). The coroutine's completion status can be checked using
+ * Done(), and for value-returning coroutines, the result can be retrieved using
+ * Value().
+ *
+ * Key features:
+ * - Support for void and value-returning coroutines.
+ * - Manual resume control via Resume().
+ * - Status checking via Done().
+ * - Value retrieval via Value() for non-void coroutines.
+ * - Awaitable interface for use in other coroutines, enabling cascading
+ * coroutines.
+ *
+ * @tparam T The type of value returned by the coroutine. Use void for
+ * coroutines that don't return a value.
+ */
+template <typename T = void>
+class Coroutine {
+  // The Promise class is used to manage the coroutine's state and control flow.
+  // We need one to handle void coroutines and another for value-returning
+  // coroutines. Their implementations are very similar, but the value-returning
+  // version stores the return value. The names of the methods are based on the
+  // C++20 coroutine specification, and cannot be changed. The two classes
+  // eventually coalesce into the Promise type alias below. The Promise class is
+  // completely internal to the Coroutine class and is not meant to be used
+  // directly.
+  struct PromiseVoid {
+    Coroutine<> get_return_object()
+    {
+      return Coroutine<>{
+          std::move(std::coroutine_handle<Promise>::from_promise(*this))};
+    }
+    std::suspend_always initial_suspend()
+    {
+      suspended_ = true;
+      return {};
+    }
+    std::suspend_always final_suspend() noexcept { return {}; }
+    void unhandled_exception() {}
+    void return_void()
+    {
+      auto awaitingCoroutine = awaitingCoroutine_;
+      if (awaitingCoroutine) {
+        awaitingCoroutine_ = nullptr;
+        std::coroutine_handle<>::from_address(awaitingCoroutine).resume();
+      }
+    }
+    [[no_unique_address]] std::monostate value_{};
```

Comment:
Is this necessary for this?:

https://github.com/triton-inference-server/perf_analyzer/pull/243/files#diff-053471f1e45d31573787f7b9b3b13d623b2d2cbff9026b8e8d43bf3d4c95f2fdR122-R124

In other words if the structs didn't have the same list of data member identifiers, the typedef line wouldn't work?

Reply:
It's for the await_resume at the end, and the various other generic places where we haul the result of the promise.

Reply:
Why not just one promise_type struct that has both return_void and return_value?

Reply:
So this is a bit on the C++ design here. There's a sort of flaw: when the language encounters `co_return`, it looks up the contents of the promise object associated with the coroutine, and makes a decision based on the presence of either `return_value` and `return_void`. To solve this connundrum, we create two promise types, and then we erase one of the two using `std::conditional` to pick the proper one. See https://en.cppreference.com/w/cpp/language/coroutines for some more details on the execution flow.

Basically, having both creates this ambiguity you shouldn't have in normal C++:
```c++
int someFunction() {
    if (someErrorCondition) {
        return;
    }
    return 42;
}
```